EventGenerator/src/CORSIKAEventGenerator_module.cc

```diff
@@ -0,0 +1,77 @@
+// Mu2e includes.
+#include "ConfigTools/inc/SimpleConfig.hh"
+#include "ConfigTools/inc/requireUniqueKey.hh"
+#include "MCDataProducts/inc/GenId.hh"
+#include "MCDataProducts/inc/GenParticleCollection.hh"
+#include "MCDataProducts/inc/G4BeamlineInfoCollection.hh"
+#include "SeedService/inc/SeedService.hh"
+
+// Includes from art and its toolchain.
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "art/Framework/Principal/Handle.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+// C++ includes.
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <memory>
+
+#include "EventGenerator/inc/CosmicCORSIKA.hh"
+
+    namespace mu2e {
+
+  class CorsikaEventGenerator : public art::EDProducer {
```

Comment:
I am not sure I understand what you mean by source module, could you please clarify?

Reply:
Instead of inheriting from EDAnalyzer or EDProducer or EDFilter you do something else ( the rules are a little more complicated).  Then you specify your module in the source : {} table of the fcl file, instead of EmptyEvent or RootInput.  The big win is that if you run out of events then you can gracefully shutdown the job.  If you implement jobs as EmptySource and a producer module then, if you run out of events, you have to throw an exception, which works but is not very user friendly.

Andrei has experience writing source modules.  You can anexamples in:
Sources/src/FromEMFMARSFileWeighted_source.cc

The interface to art is not via inheritance.  Instead you write a "detail class" then the actual source module is the instantiation of a source module template using your detail class as the template argument.   It's just a different way for you to provide callbacks that are called by the art code at the right time with the right arguments.

Reply:
I am afraid this is beyond my C++ competencies... I left it as it was, also because it is the same approach of CRY, but I am happy to change it with a little of help from more expert people. Also, at the moment the module can't run out of events because if it reaches the last binary file it starts again from the first one, shifting the position of the cosmic rays by a random amount.

Reply:
Ok, I talked with @brownd1978 and I see your point now. I will work on it but it will take a couple of days since I am not very familiar with it. At the moment I was avoiding the "end of the file" problem by starting again from the first file and moving the particle by a random amount (which is basically resampling) but I see it's better to separate the two logics.

Reply:
I wrote the Source module, however it wasn't as straightforward as I thought: the source module could access the geometry only after the first event had been processed, so I couldn't extrapolate the cosmic rays to the world borders. In order to avoid ugly solutions (such as storing a first empty event) I had to write also a producer which takes as input the output of the Source module and does the backwards extrapolation. I compared the outputs before and after these changes and they are the same.